### PR TITLE
[WIP] Changing nlp from dict to class

### PR DESCRIPTION
**Pull request by Steakkk**
### All Submissions:

* [ ] Have you followed the guidelines in our Contributing document [docs/contribution.md]?
* [ ] Have you checked to ensure there aren't other open [Pull Requests] for the same update/change?
* [ ] Have you opened/linked the issue related to your pull request?
* [ ] Have you used the tag [WIP] for on-going changes, and removed it when the pull request was ready?
* [ ] When ready to merge, have you sent a comment pinging @pariterre in it?

### New Feature Submissions:

1. [ ] Does your submission pass the tests (if not please explain why this is intended)?
2. [ ] Have you linted your code locally prior to submission (using the command: `black . -l120 --exclude "external/*"`)?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new examples for your core changes, as applicable?
* [ ] Have you written new tests for your core changes, as applicable?

### CODE REVIEW

File: biorbd_optim/dynamics/dynamics_functions.py [hunk 1]
```diff
@@ -25,18 +25,18 @@ def forward_dynamics_torque_driven(states, controls, parameters, nlp):
...
-        if "external_forces" in nlp:
-            dxdt = MX(nlp["nx"], nlp["ns"])
-            for i, f_ext in enumerate(nlp["external_forces"]):
-                qddot = nlp["model"].ForwardDynamics(q, qdot, tau, f_ext).to_mx()
-                qddot_reduced = nlp["q_dot_mapping"].reduce.map(qddot)
+        if nlp.external_forces:
+            dxdt = MX(nlp.nx, nlp.ns)
+            for i, f_ext in enumerate(nlp.external_forces):
+                qddot = nlp.model.ForwardDynamics(q, qdot, tau, f_ext).to_mx()
+                qddot_reduced = nlp.q_dot_mapping.reduce.map(qddot)
```
Comment (pariterre) [inline biorbd_optim/dynamics/dynamics_functions.py:35]: for mapping use nlp.mapping["q_dot"]

File: biorbd_optim/dynamics/dynamics_functions.py [hunk 2]
```diff
@@ -341,18 +341,18 @@ def forces_from_forward_dynamics_muscle_excitations_and_torque_driven_with_conta
         DynamicsFunctions.apply_parameters(parameters, nlp)
         q, qdot, residual_tau = DynamicsFunctions.dispatch_q_qdot_tau_data(states, controls, nlp)
 
-        muscles_states = biorbd.VecBiorbdMuscleState(nlp["nbMuscle"])
-        muscles_excitation = controls[nlp["nbTau"] :]
-        muscles_activations = states[nlp["nbQ"] + nlp["nbQdot"] :]
+        muscles_states = biorbd.VecBiorbdMuscleState(nlp.nbMuscle)
```
Comment (pariterre) [inline biorbd_optim/dynamics/dynamics_functions.py:344]: Replace nbX by nlp.shape["muscles"]

File: biorbd_optim/dynamics/problem.py [hunk 3]
```diff
@@ -74,8 +74,8 @@ def torque_activations_driven_with_contact(ocp, nlp):
         """
         Problem.configure_q_qdot(nlp, True, False)
         Problem.configure_tau(nlp, False, True)
-        nlp["nbActuators"] = nlp["nbTau"]
-        if nlp["dynamics_type"].dynamics:
+        nlp.nbActuators = nlp.nbTau
+        if nlp.dynamics_type.dynamics:
```
Comment (pariterre) [inline biorbd_optim/dynamics/problem.py:78]: Please have a look if `dynamics_func` and `dynamics` are the same. If so, remove `nlp.dynamics_type.dynamics`

File: biorbd_optim/limits/constraints.py [hunk 4]
```diff
@@ -106,7 +106,7 @@ def non_slipping(
 
             mu = static_friction_coefficient
             for i in range(len(u)):
-                contact = nlp["contact_forces_func"](x[i], u[i], p)
+                contact = nlp.contact_forces_func(x[i], u[i], p)
```
Comment (pariterre) [inline biorbd_optim/limits/constraints.py:109]: `contact_forces_func` should be somehow move next to `dynamics`  (or `dynamics_func`)

**Review by pariterre — CHANGES_REQUESTED** [review]

File: biorbd_optim/gui/plot.py [hunk 5]
```diff
@@ -57,7 +57,7 @@ class PlotOcp:
     def __init__(self, ocp, automatically_organize=True, adapt_graph_size_to_bounds=False):
         """Prepares the figure"""
         for i in range(1, ocp.nb_phases):
-            if ocp.nlp[0]["nbQ"] != ocp.nlp[i]["nbQ"]:
+            if ocp.nlp[0].nbQ != ocp.nlp[i].nbQ:
```
Comment (pariterre) [inline biorbd_optim/gui/plot.py:60]: This should be shape["Q"]

**Review by pariterre — CHANGES_REQUESTED** [review]

File: biorbd_optim/dynamics/problem.py (hunk 3 above)
Comment (Steakkk) [reply biorbd_optim/dynamics/problem.py:78]: It looks like these variables are not the same, I found three different places where they don't have the same value

File: biorbd_optim/misc/non_linear_program.py [hunk 6]
```diff
@@ -55,19 +71,27 @@ def __init__(
...
         self.ode_solver = ode_solver
         self.p = p
         self.par_dynamics = par_dynamics
+        self.parameters_to_optimize = parameters_to_optimize
         self.phase_idx = phase_idx
         self.plot = plot
-        self.shape = (shape,)
+        self.problem_type = problem_type
+        self.q = q
+        self.qdot = qdot
```
Comment (pariterre) [inline biorbd_optim/misc/non_linear_program.py:92]: For homogeonity sake, can you make it q_dot

File: biorbd_optim/misc/optimal_control_program.py [hunk 7]
```diff
@@ -263,16 +263,17 @@ def __init__(
             if q_mapping is not None or q_dot_mapping is not None or tau_mapping is not None:
                 raise RuntimeError("all_generalized_mapping and a specified mapping cannot be used alongside")
             q_mapping = q_dot_mapping = tau_mapping = all_generalized_mapping
-        self.__add_to_nlp("q_mapping", q_mapping, q_mapping is None, BidirectionalMapping)
-        self.__add_to_nlp("q_dot_mapping", q_dot_mapping, q_dot_mapping is None, BidirectionalMapping)
-        self.__add_to_nlp("tau_mapping", tau_mapping, tau_mapping is None, BidirectionalMapping)
+        # TODO: find a clean solution for those
+        self.__add_to_nlp("q", q_mapping, q_mapping is None, BidirectionalMapping, mapping=True)
```
Comment (pariterre) [inline biorbd_optim/misc/optimal_control_program.py:267]: I think using a flag for the entry is a good idea, but mapping is a bit too specific. I suggest that the value is a None by default and is therefore added as a "name" attribute, or the value is a string and is thereafter added to a dictionary

**Review by pariterre — CHANGES_REQUESTED** [review]
It's getting there :)

**Review by pariterre — COMMENTED** [review]
Keeping up

File: biorbd_optim/misc/optimal_control_program.py (hunk 7 above)
Comment (Steakkk) [reply biorbd_optim/misc/optimal_control_program.py]: Done! I'm not sure I did exactly what you wanted, now mapping is a string named name, with default value None, and if it is not None I just add param to the dictionnary designed by name

File: biorbd_optim/limits/constraints.py (hunk 4 above)
Comment (pariterre) [reply biorbd_optim/limits/constraints.py:109]: Don't loose time on time, it is okay as is

**Review by pariterre — APPROVED** [review]

**Comment by pariterre** [comment]
LGTM!
GG

**Pull request merged**